datafusion/physical-plan/src/sorts/partial_sort.rs

```diff
@@ -0,0 +1,992 @@
+    ///
+    /// Return None if the batch cannot be partitioned, which means the
+    /// batch does not have the information for a safe sort
+    fn get_slice_point(
+        &self,
+        common_prefix_len: usize,
+        batch: &RecordBatch,
+    ) -> Result<Option<usize>> {
+        let common_prefix_sort_keys = (0..common_prefix_len)
+            .map(|idx| self.expr[idx].evaluate_to_sort_column(batch))
+            .collect::<Result<Vec<_>>>()?;
+        let partition_points =
+            evaluate_partition_ranges(batch.num_rows(), &common_prefix_sort_keys)?;
+        // If partition points are [0..100], [100..200], [200..300]
+        // we should return 200, which is the safest and furthest partition boundary
+        // Please note that we shouldn't return 300 (which is number of rows in the batch),
+        // because this boundary may change with new data.
+        if partition_points.len() >= 2 {
+            Ok(Some(partition_points[partition_points.len() - 2].end))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use arrow::array::*;
+    use arrow::compute::SortOptions;
+    use arrow::datatypes::*;
+    use futures::FutureExt;
+    use itertools::Itertools;
+
+    use datafusion_common::assert_batches_eq;
+
+    use crate::collect;
+    use crate::expressions::col;
+    use crate::memory::MemoryExec;
+    use crate::sorts::sort::SortExec;
+    use crate::test;
+    use crate::test::assert_is_pending;
+    use crate::test::exec::{assert_strong_count_converges_to_zero, BlockingExec};
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_partial_sort() -> Result<()> {
+        let task_ctx = Arc::new(TaskContext::default());
+        let source = test::build_table_scan_i32(
+            ("a", &vec![0, 0, 0, 1, 1, 1]),
+            ("b", &vec![1, 1, 2, 2, 3, 3]),
+            ("c", &vec![1, 0, 5, 4, 3, 2]),
+        );
+        let schema = Schema::new(vec![
+            Field::new("a", DataType::Int32, false),
+            Field::new("b", DataType::Int32, false),
+            Field::new("c", DataType::Int32, false),
+        ]);
+        let option_asc = SortOptions {
+            descending: false,
+            nulls_first: false,
+        };
+
+        let partial_sort_exec = Arc::new(
+            PartialSortExec::new(
+                vec![
+                    PhysicalSortExpr {
+                        expr: col("a", &schema)?,
+                        options: option_asc,
+                    },
+                    PhysicalSortExpr {
+                        expr: col("b", &schema)?,
+                        options: option_asc,
+                    },
+                    PhysicalSortExpr {
+                        expr: col("c", &schema)?,
+                        options: option_asc,
+                    },
+                ],
+                source.clone(),
+            )
+            .with_common_prefix_length(2),
```

Comment:
I recommend extending this test to also use a common prefix length of 1 (which should still produce the same answer)